util/network-devp2p/src/host.rs

```diff
@@ -908,13 +903,11 @@ impl Host {
 
 	fn connection_timeout(&self, token: StreamToken, io: &IoContext<NetworkIoMessage>) {
 		trace!(target: "network", "Connection timeout: {}", token);
-		self.kill_connection(token, io, true)
+		self.kill_connection(token, io)
 	}
 
-	fn kill_connection(&self, token: StreamToken, io: &IoContext<NetworkIoMessage>, remote: bool) {
+	fn kill_connection(&self, token: StreamToken, io: &IoContext<NetworkIoMessage>) {
```

Comment:
So now we are `deregistering` the stream everytime we call `kill_connection`. Previously it was done only if `session` was done. Seems that `deregister` will remove entry from `self.sessions`, but only when session is expired - otherwise we will leak the the value in the `HashMap` forever.

The code is really fragile here, and I think we should be careful with removing stuff that just "seems used incorrectly", especially when it runs for couple of years now and we are not really sure if there are any bugs in that code (i.e. is this strictly related to TCP streams hanging in the WAIT state?).

Reply:
Second issue: So now we don't `note_failure` on any of the previous `kill_connection(_,_, true)` - to avoid breaking stuff we should still review all call sites of that function if `note_failure` should be there.

Reply:
Yeah, seems strange to remove `remote` just because it was wrongly used. Maybe just rename the variable? Or have a `deregister` bool and a `note_failure` one.

Reply:
> So now we are deregistering the stream every time we call `kill_connection`. Previously it was done only if session was done. Seems that deregister will remove entry from self.sessions, but only when session is expired - otherwise we will leak the the value in the HashMap forever.

@tomusdrw the leak is happening now. Session is always set to expired, `session` is removed from handlers, but the value is never deregistered if `remote = false`, cause `s.done()` returns true as long as the socket is `writeable`.

In this pr the leak is not happening, cause after setting expired to true, and unregistering handlers, we always call deregister.

https://github.com/paritytech/parity-ethereum/blob/05550429c7dc0e154710d171bb432e68d000e34d/util/network-devp2p/src/host.rs#L917-L926

> The code is really fragile here, and I think we should be careful with removing stuff that just "seems used incorrectly", especially when it runs for couple of years now and we are not really sure if there are any bugs in that code (i.e. is this strictly related to TCP streams hanging in the WAIT state?).

I believe that `CLOSE_WAIT` state is just a symptom of a bigger problem. After we call `kill_connection`, sessions leak in memory and they are already disconnected from the handlers. Remote nodes close the connection to our unresponsive socket, but because they are no handlers connected, we are stuck on the `CLOSE_WAIT` state.